// @adobe/ccweb-add-on-analytics/dist/base/BaseCommand.js
/********************************************************************************
 * MIT License

 * © Copyright 2025 Adobe. All rights reserved.

 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 ********************************************************************************/
import { Command, Flags } from "@oclif/core";
import { IContainer, ITypes } from "../config/index.js";
export class BaseCommand extends Command {
    _analyticsConsent;
    _analyticsService;
    constructor(argv, config, program) {
        super(argv, config);
        this._analyticsConsent = IContainer.get(ITypes.AnalyticsConsent);
        this._analyticsService = IContainer.get(ITypes.AnalyticsService);
        this._analyticsService.program = program;
        this._analyticsService.startTime = Date.now();
    }
    static baseFlags = {
        analytics: Flags.string({
            char: "a",
            description: "Turn on/off sending analytics to Adobe.",
            options: ["on", "off"],
            required: false
        }),
        verbose: Flags.boolean({
            char: "v",
            description: "Enable verbose logging.",
            default: false,
            required: false
        })
    };
    async _seekAnalyticsConsent(analytics) {
        if (analytics === undefined) {
            await this._analyticsConsent.get();
        }
        else {
            await this._analyticsConsent.set(analytics === "on");
        }
    }
}
